resolwe/observers/models.py

```diff
@@ -0,0 +1,165 @@
+"""The model Observer model."""
+import uuid
+from typing import List, Optional
+
+from asgiref.sync import async_to_sync
+from channels.layers import get_channel_layer
+
+from django.conf import settings
+from django.contrib.contenttypes.models import ContentType
+from django.db import models, transaction
+from django.db.models import Count, Q
+from django.db.models.query import QuerySet
+
+from resolwe.permissions.models import Permission
+
+from .protocol import (
+    CHANGE_TYPE_CREATE,
+    CHANGE_TYPE_DELETE,
+    CHANGE_TYPE_UPDATE,
+    GROUP_SESSIONS,
+    TYPE_ITEM_UPDATE,
+)
+
+
+def get_random_uuid() -> str:
+    """Generate a random UUID in string format."""
+    return uuid.uuid4().hex
+
+
+class Observer(models.Model):
+    """An observer for detecting changes in a model.
+
+    Characterized by a db table or an instance of a model and a change type.
+    Several subscriptions can subscribe to the same observer.
+    """
+
+    CHANGE_TYPES = (
+        (CHANGE_TYPE_CREATE, "create"),
+        (CHANGE_TYPE_UPDATE, "update"),
+        (CHANGE_TYPE_DELETE, "delete"),
+    )
+
+    #: Table of the observed resource.
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    #: Primary key of the observed resource (null if watching the whole table).
+    object_id = models.PositiveIntegerField(null=True)
+    #: The type of change to observe for.
+    change_type = models.CharField(choices=CHANGE_TYPES, max_length=6)
+
+    class Meta:
+        """Meta parameters for the Observer model."""
+
+        unique_together = ("content_type", "object_id", "change_type")
+        indexes = [models.Index(fields=["object_id"])]
+
+    @classmethod
+    def get_interested(
+        cls,
+        content_type: ContentType,
+        object_id: Optional[int] = None,
+        change_type: Optional[str] = None,
+    ) -> "QuerySet[Observer]":
+        """Find all observers watching for changes of a given item/table."""
+        query = Q(content_type=content_type)
+        if change_type is not None:
+            query &= Q(change_type=change_type)
+        query &= Q(object_id=object_id) | Q(object_id__isnull=True)
+        return cls.objects.filter(query)
+
+    @classmethod
+    def observe_instance_changes(cls, instance: any, change_type: str):
```

Comment:
You probably could use at least ``django.db.models.Model`` for the instance base class.